accepted/2022/simplify-rid-model.md

```diff
@@ -0,0 +1,240 @@
+
+For .NET 7:
+
+- For Windows, .NET will target the Windows 10 CRT.
+- For Linux with glibc, .NET will target CentOS 7.
+- For Linux with musl, .NET will target the oldest supported Alpine version.
+
+For .NET 8, we'll continue with the model. However, we will no longer be able to use [CentOS 7 (due to EOL)](https://wiki.centos.org/About/Product), but will need to adopt another distro that provides us with an old enough glibc version such that we can offer broad compatibility.
+
+For .NET 9, we'll likely drop support for Windows 10 and instead target the Windows 11 CRT.
+
+As part of investigating this topic, we noticed the [libc compatibility model that TensorFlow uses](https://github.com/tensorflow/tensorflow/blob/f3963e82f21c9d094503568699877655f9dac508/tensorflow/tools/ci_build/devtoolset/build_devtoolset.sh#L48-L57). That model enables the use of a modern OS with an artificially old libc. We also saw that Python folks are doing something similar with their [`manylinux` approach but with docker](https://github.com/pypa/manylinux).
+
+This in turn made us realize that our [`build-tools-prereqs` Docker images](https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/main/src/centos/7/Dockerfile) are very similar to the Python approach. We don't need a new contract like `manylinux2014` since we can rely on the contract changing with each major .NET version. We need augment our build-tools-prereq scheme to also include an old glibc version, once CentOS 7 is EOL.
+
+The biggest difference with our `build-tools-prereqs` images is that they are not intended for others to use, while the `manylinux` images are intended as a general community artifact. There are two primary cases where extending the use of the `build-tools-prereqs` images would be useful: enabling others to produce a .NET build with the same compatibility reach, and enabling NuGet authors to build native dependencies with matching compatibility as the underlying runtime. Addressing that topic is outside the scope of this document. This context is included to help frame how the TensorFlow, Python, and .NET solutions compare, and to inspire future conversation. 
+
+Note: It appears that some other folks have been [reacting to CentOS 7 not being a viable compilation target](https://gist.github.com/wagenet/35adca1a032cec2999d47b6c40aa45b1) for much longer.
+
+## Legal RIDs
+
+RIDs are defined by legal combinations of operating system, chip architecture, and C-runtime. We support both triplets (primarily native code) and singles (only processor-agnostic managed code).
+
+A given RID is used to describe:
+
+- This system supports this RID, per the .NET host.
+- This code will run on / requires a system that supports this RID, per the author.
+
+The following table describes RIDs supported by the .NET host. It is exhaustive, per the hosts provided by Microsoft.
```

Comment:
How do we deal with architectures like `s390x` that Microsoft doesn't support and/or provide builds for?

Reply:
That's what this is about: https://github.com/dotnet/designs/pull/260/files#diff-20aa11d3c5c8d54fc1c45f4b81927255d3951fdd7cf58589fc6d03a2dd2a6a1bR221

The proposal is to NOT add any new RIDs, but enable source-builds to add additive `runtimes.json` files that augment the graph. So, frozen is kind a soft term. It means that we want to avoid creating any new official RIDs, but that distro maintainers can create their own for their own distro. The assumption is that they are only needed to create a link between the user and the runtime/host packs and not NuGet. As a result, these RIDs don't need to be stored anywhere else. We'll accept a file like `runtimes-s390x.json` in upstream. It would probably be <20 lines.

Fair?

Reply:
> The assumption is that they are only needed to create a link between the user and the runtime/host packs and not NuGet

In the SkiSharp scenario, the authors of SkiaSharp might want to publish `linux-s390x` assets on nuget.org. What would that look like?

Alternatively, do you think it might be worth it to blanket-add all the current architectures supported by mono to runtime.json before freezing it?

Reply:
> In the SkiSharp scenario, the authors of SkiaSharp might want to publish linux-s390x assets on nuget.org. What would that look like?

They can add an so file in their NuGet package at runtimes/linux-s390x, and then the package will support s390x.

```
$ wget --no-check-certificate https://www.nuget.org/api/v2/package/SkiaSharp.NativeAssets.Linux/2.88.0-preview.232
$ unzip 2.88.0-preview.232
$ cd runtimes/
$ tree
.
├── linux-arm
│   └── native
│       └── libSkiaSharp.so
├── linux-arm64
│   └── native
│       └── libSkiaSharp.so
├── linux-musl-x64
│   └── native
│       └── libSkiaSharp.so
└── linux-x64
    └── native
        └── libSkiaSharp.so
```

The `linux-s390x` rid could be added as part of source-build which means it will be usable on the resulting .NET distribution.

The reason for adding it to upstream `runtimes.json` is to make it a known rid for to **all** .NET distributions. Then they can publish against that rid. Otherwise it would be unknown.

Reply:
That's a good point. Agree. Is that good @omajid? If so, I'll update with the spec with the wording from @tmds.

Reply:
Updated the doc.